### PR TITLE
fix(manager/composer): respect the PHP override in composer.json

### PR DESCRIPTION
**Pull request by internalsystemerror**
## Changes:

This ensures that renovate respects the PHP override set in `composer.json`: https://getcomposer.org/doc/06-config.md#platform.

Given the following:

```json
{
    "config": {
        "platform": {
            "php": "7.4.27"
        }
    },
    "require": {
        "php": "^7.4 || ~8.0.0"
    }
}
```

Then the detected PHP version used for composer should be `7.4.27`. This PR adds that awareness to renovate.

## Context:

There are several issues related which should be closed by this (for instance https://github.com/renovatebot/renovate/issues/4396). And mainly came about after this discussion https://github.com/renovatebot/renovate/discussions/13637.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/manager/composer/types.ts [hunk 1]
```diff
@@ -8,6 +8,11 @@ export interface Repo {
 }
 export interface ComposerConfig {
   type?: string;
+  config?: {
```
Comment (viceice) [inline lib/manager/composer/types.ts:11]: Please add tsdoc with a link to the composer docs.

File: lib/manager/composer/types.ts [hunk 2]
```diff
@@ -8,6 +8,11 @@ export interface Repo {
...
+    platform?: {
+      php?: string;
```
Comment (viceice) [inline lib/manager/composer/types.ts:13]: It looks like renovate should also be able to suggest updated to this config? At least security updates.

**Review by viceice — COMMENTED** [review]
Otherwise LGTM

File: lib/manager/composer/types.ts (hunk 2 above)
Comment (internalsystemerror) [reply lib/manager/composer/types.ts:13]: I agree that would be very useful, I'm unsure how that would be implemented though.

Comment (viceice) [reply lib/manager/composer/types.ts:19]: No problem, out of scope for this PR, but we should open a follow up issue

Comment (internalsystemerror) [reply lib/manager/composer/types.ts:19]: I'll make a note to do so, thanks for the suggestion!

File: lib/manager/composer/types.ts (hunk 1 above)
Comment (internalsystemerror) [reply lib/manager/composer/types.ts:17]: done

File: lib/manager/composer/utils.ts [hunk 3]
```diff
@@ -69,7 +69,9 @@ export function extractContraints(
   const res: Record<string, string> = { composer: '1.*' };
 
   // extract php
-  if (composerJson.require?.php) {
+  if (composerJson.config?.platform?.php) {
+    res.php = composerJson.config?.platform?.php;
```
Comment (rarkins) [inline lib/manager/composer/utils.ts:73]:
```suggestion
    res.php = composerJson.config.platform.php;
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/manager/composer/types.ts (hunk 1 above)
Comment (viceice) [reply lib/manager/composer/types.ts:17]: > done

https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#resolve-review-comments-instead-of-commenting

😉

File: lib/manager/composer/utils.ts [hunk 4]
```diff
@@ -69,7 +69,9 @@ export function extractContraints(
...
+  } else if (composerJson.require?.php) {
     res.php = composerJson.require?.php;
```
Comment (viceice) [inline lib/manager/composer/utils.ts:75]:
```suggestion
     res.php = composerJson.require.php;
```
Same here. 😅

File: lib/manager/composer/utils.ts (hunk 3 above)
Comment (viceice) [reply lib/manager/composer/utils.ts:73]: Yes, optional chain operator is not required here, as we validate in if.

**Comment by internalsystemerror** [comment]
I will be able to fix this and finish this off tonight

File: lib/manager/composer/utils.ts (hunk 3 above)
Comment (internalsystemerror) [reply lib/manager/composer/utils.ts]: These are all over this file btw, you may wish to enforce this using a lint rule if possible? Do you want me to fix the other lines too? Such as:
```ts
if (composerJson.require?.['composer/composer']) {
    res.composer = composerJson.require?.['composer/composer'];
  } else if (composerJson['require-dev']?.['composer/composer']) {
    res.composer = composerJson['require-dev']?.['composer/composer'];
  }
```

**Comment by internalsystemerror** [comment]
Never mind, that was a quick fix. Sorry I'd missed this on the initial upload so I was able to push it now.

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Comment by internalsystemerror** [comment]
@rarkins @viceice Thank you both for your help with this!

**Pull request merged**